### DISCUSSION TITLE
React to image change events from registries

### DISCUSSION THREAD

**Discussion by bjartek** [fluxcd/flux2#693] [Proposals] [answered]
Have you considered implementing a feature where webhook/events from ImageRegistries can be used instead of polling?

There will then need to be a component that will listen to these events and then update relevant files in the repos and then push it?

**Comment by phillebaba** [comment]
Your best option there is to do what I have been doing even back in FluxV1. Do it in CI rather than in the cluster. There are multiple ways to trigger a CI pipeline in most git providers when a new image is pushed, or you could have one pipeline trigger another if that is not possible. Then in the pipeline you can build whatever custom logic you want to update the image tag and/or do promotion to other environments.

**Reply by stefanprodan** [reply]
I think you got the question wrong. I agree with @bjartek that we should react to image push events the same way we do for git push. Instead of waiting for the image scan interval, the event should trigger it instead.

**Comment by stefanprodan** [comment, marked as answer]
My proposal is to extend the notification-controller [Receiver API](https://toolkit.fluxcd.io/components/notification/receiver/) to  trigger an `ImageRepository` reconciliation:
- the generic receiver could be used from CI to trigger an image scan+update 
- allow the Harbor and GitHub (GHCR) receivers to trigger an image scan+update 
- implement Docker Hub and Quay receivers (maybe others?)

**Reply by phillebaba** [reply]
Yeah I guess I understood the question wrong in that case. Sure this sounds like a good idea. As long as it is reasonably documented implementing it should not be a problem.

**Reply by bjartek** [reply]
Sounds like a very nice feature @stefanprodan

**Comment by stefanprodan** [comment]
Triggering image updates from CI/Harbor/GitHub/GitLab/BitBucket implemented in https://github.com/fluxcd/notification-controller/pull/110

**Reply by bjartek** [reply]
Awesome!

**Comment by stefanprodan** [comment]
Container registry integrations:

- [x] DockerHub https://github.com/fluxcd/notification-controller/pull/112
- [x] Harbor https://github.com/fluxcd/notification-controller/pull/110
- [ ] Quay https://github.com/fluxcd/notification-controller/issues/114
- [ ] Nexus https://github.com/fluxcd/notification-controller/issues/115
- [ ] GCR https://github.com/fluxcd/notification-controller/issues/116
- [ ] ACR https://github.com/fluxcd/notification-controller/issues/117

@SomtochiAma will be working on getting all of these done.
